src/hotspot/share/gc/g1/g1ConcurrentMarkThread.cpp

```diff
@@ -134,14 +160,23 @@ class G1ConcPhaseTimer : public GCTraceConcTimeImpl<LogLevel::Info, LOG_TAGS(gc,
 void G1ConcurrentMarkThread::run_service() {
   _vtime_start = os::elapsedVTime();
 
-  while (wait_for_next_cycle()) {
+  Command cmd;
+  while ((cmd = wait_for_next_cycle()) != Terminate) {
 
     GCIdMark gc_id_mark;
-    GCTraceConcTime(Info, gc) tt("Concurrent Cycle");
+    GCTraceConcTime(Info, gc) tt(FormatBuffer<128>("Concurrent %s Cycle",
+                                                   cmd == MarkCycle ? "Mark" : "Undo"));
 
     concurrent_cycle_start();
-    full_concurrent_cycle_do();
-    concurrent_cycle_end();
+
+    if (cmd == MarkCycle) {
+      concurrent_mark_cycle_do();
+    } else {
+      assert(cmd == UndoCycle, "Must be command to undo concurrent start but is %d", cmd);
+      concurrent_undo_cycle_do();
+    }
+
+    concurrent_cycle_end(cmd == MarkCycle && !_cm->has_aborted());
```

Comment:
I think it would be more clear if this was called with `false` for the undo-case and with `!_cm->has_aborted()` for the normal case. But if you want to avoid two calls, I'm fine with this.